File: app/routers/delegate_broadcast.py
"""Broadcast delegation router (DELEGATE-004).

Endpoints for delegated broadcast chat moderation (pin, delete, mute, ban,
announce) and broadcast control (start, stop, schedule).
"""

from __future__ import annotations

from typing import List

from fastapi import APIRouter, Depends, Query

from app.models import (
    BroadcastMuteIn,
    BroadcastBanIn,
    BroadcastAnnouncementIn,
    BroadcastScheduleIn,
    BroadcastModeratorOut,
    BroadcastBanOut,
    BroadcastModerationLogEntry,
)
from app.services import delegate_broadcast as svc
from app.services.sessions import require_ui_session

router = APIRouter(prefix="/ui/broadcast/delegate", tags=["broadcast-delegation"])


# ---------------------------------------------------------------------------
# Chat moderation
# ---------------------------------------------------------------------------


@router.post("/{creator_id}/sessions/{sid}/chat/{mid}/pin")
async def pin_message(
    creator_id: str,
    sid: str,
    mid: str,
    user=Depends(require_ui_session),
):
    """Pin a chat message in a broadcast session."""
    return svc.pin_chat_message(
        session_id=sid,
        message_id=mid,
        moderator_id=user["user_sub"],
        creator_id=creator_id,
    )


@router.delete("/{creator_id}/sessions/{sid}/chat/{mid}/pin")
async def unpin_message(
    creator_id: str,
    sid: str,
    mid: str,
    user=Depends(require_ui_session),
):
    """Unpin a chat message in a broadcast session."""
    return svc.unpin_chat_message(
        session_id=sid,
        message_id=mid,
        moderator_id=user["user_sub"],
        creator_id=creator_id,
    )


@router.delete("/{creator_id}/sessions/{sid}/chat/{mid}")
async def delete_message(
    creator_id: str,
    sid: str,
    mid: str,
    user=Depends(require_ui_session),
):
    """Delete a chat message from a broadcast session."""
    return svc.delete_moderated_message(
        session_id=sid,
        message_id=mid,
        moderator_id=user["user_sub"],
        creator_id=creator_id,
    )


@router.post("/{creator_id}/sessions/{sid}/mute")
async def mute_viewer(
    creator_id: str,
    sid: str,
    body: BroadcastMuteIn,
    user=Depends(require_ui_session),
):
    """Mute a viewer in broadcast chat."""
    return svc.mute_viewer(
        session_id=sid,
        user_id=body.user_id,
        moderator_id=user["user_sub"],
        creator_id=creator_id,
        duration_seconds=body.duration_seconds,
    )


@router.post("/{creator_id}/sessions/{sid}/ban")
async def ban_viewer(
    creator_id: str,
    sid: str,
    body: BroadcastBanIn,
    user=Depends(require_ui_session),
):
    """Ban a viewer from broadcast chat."""
    return svc.ban_viewer(
        session_id=sid,
        user_id=body.user_id,
        moderator_id=user["user_sub"],
        creator_id=creator_id,
        reason=body.reason,
    )


@router.delete("/{creator_id}/sessions/{sid}/ban/{uid}")
async def unban_viewer(
    creator_id: str,
    sid: str,
    uid: str,
    user=Depends(require_ui_session),
):
    """Unban a viewer from broadcast chat."""
    return svc.unban_viewer(
        session_id=sid,
        user_id=uid,
        moderator_id=user["user_sub"],
        creator_id=creator_id,
    )


@router.post("/{creator_id}/sessions/{sid}/announcement")
async def post_announcement(
    creator_id: str,
    sid: str,
    body: BroadcastAnnouncementIn,
    user=Depends(require_ui_session),
):
    """Post an announcement in broadcast chat."""
    return svc.post_announcement(
        session_id=sid,
        moderator_id=user["user_sub"],
        creator_id=creator_id,
        text=body.text,
    )


# ---------------------------------------------------------------------------
# Broadcast control
# ---------------------------------------------------------------------------


@router.post("/{creator_id}/sessions/{sid}/start")
async def start_broadcast(
    creator_id: str,
    sid: str,
    user=Depends(require_ui_session),
):
    """Start a broadcast on behalf of the creator."""
    return svc.start_broadcast_as_delegate(
        session_id=sid,
        delegate_id=user["user_sub"],
        creator_id=creator_id,
    )


@router.post("/{creator_id}/sessions/{sid}/stop")
async def stop_broadcast(
    creator_id: str,
    sid: str,
    user=Depends(require_ui_session),
):
    """Stop a broadcast on behalf of the creator."""
    return svc.stop_broadcast_as_delegate(
        session_id=sid,
        delegate_id=user["user_sub"],
        creator_id=creator_id,
    )


@router.post("/{creator_id}/sessions/schedule")
async def schedule_broadcast(
    creator_id: str,
    body: BroadcastScheduleIn,
    user=Depends(require_ui_session),
):
    """Schedule a broadcast on behalf of the creator."""
    return svc.schedule_broadcast_as_delegate(
        delegate_id=user["user_sub"],
        creator_id=creator_id,
        title=body.title,
        scheduled_at=body.scheduled_at,
        profile_id=body.profile_id,
    )


# ---------------------------------------------------------------------------
# Moderator management
# ---------------------------------------------------------------------------


@router.post("/{creator_id}/sessions/{sid}/moderator/register")
async def register_moderator(
    creator_id: str,
    sid: str,
    user=Depends(require_ui_session),
):
    """Register the caller as an active moderator for a broadcast session."""
    return svc.register_moderator(
        session_id=sid,
        moderator_id=user["user_sub"],
        creator_id=creator_id,
    )


@router.get(
    "/{creator_id}/sessions/{sid}/moderators",
    response_model=List[BroadcastModeratorOut],
)
async def list_moderators(
    creator_id: str,
    sid: str,
    user=Depends(require_ui_session),
):
    """List active moderators for a broadcast session."""
    return svc.list_active_moderators(sid)


@router.get(
    "/{creator_id}/sessions/{sid}/bans",
    response_model=List[BroadcastBanOut],
)
async def list_bans(
    creator_id: str,
    sid: str,
    user=Depends(require_ui_session),
):
    """List banned viewers for a broadcast session."""
    return svc.list_banned_viewers(sid)


@router.get(
    "/{creator_id}/sessions/{sid}/moderation-log",
    response_model=List[BroadcastModerationLogEntry],
)
async def get_moderation_log(
    creator_id: str,
    sid: str,
    limit: int = Query(default=100, ge=1, le=500),
    user=Depends(require_ui_session),
):
    """Get moderation action log for a broadcast session."""
    return svc.get_broadcast_moderation_log(sid, limit=limit)

File: frontend/src/api/endpoints/delegateBroadcast.ts
import { api } from "@/api/client";
import type {
  BroadcastMuteReq,
  BroadcastBanReq,
  BroadcastAnnouncementReq,
  BroadcastScheduleReq,
  BroadcastModeratorOut,
  BroadcastBanOut,
  BroadcastModerationLogEntry,
} from "@/api/types";

const BASE = "/ui/broadcast/delegate";

// -- Chat moderation --

export async function pinMessage(
  creatorId: string,
  sessionId: string,
  messageId: string,
): Promise<{ ok: boolean; message_id: string; pinned: boolean }> {
  return api.post(`${BASE}/${creatorId}/sessions/${sessionId}/chat/${messageId}/pin`, {});
}

export async function unpinMessage(
  creatorId: string,
  sessionId: string,
  messageId: string,
): Promise<{ ok: boolean; message_id: string; pinned: boolean }> {
  return api.del(`${BASE}/${creatorId}/sessions/${sessionId}/chat/${messageId}/pin`);
}

export async function deleteMessage(
  creatorId: string,
  sessionId: string,
  messageId: string,
): Promise<{ ok: boolean; message_id: string }> {
  return api.del(`${BASE}/${creatorId}/sessions/${sessionId}/chat/${messageId}`);
}

export async function muteViewer(
  creatorId: string,
  sessionId: string,
  req: BroadcastMuteReq,
): Promise<Record<string, unknown>> {
  return api.post(`${BASE}/${creatorId}/sessions/${sessionId}/mute`, req);
}

export async function banViewer(
  creatorId: string,
  sessionId: string,
  req: BroadcastBanReq,
): Promise<Record<string, unknown>> {
  return api.post(`${BASE}/${creatorId}/sessions/${sessionId}/ban`, req);
}

export async function unbanViewer(
  creatorId: string,
  sessionId: string,
  userId: string,
): Promise<{ ok: boolean }> {
  return api.del(`${BASE}/${creatorId}/sessions/${sessionId}/ban/${userId}`);
}

export async function postAnnouncement(
  creatorId: string,
  sessionId: string,
  req: BroadcastAnnouncementReq,
): Promise<{ ok: boolean; message_id: string; is_announcement: boolean }> {
  return api.post(`${BASE}/${creatorId}/sessions/${sessionId}/announcement`, req);
}

// -- Broadcast control --

export async function startBroadcast(
  creatorId: string,
  sessionId: string,
): Promise<{ ok: boolean; session_id: string; status: string }> {
  return api.post(`${BASE}/${creatorId}/sessions/${sessionId}/start`, {});
}

export async function stopBroadcast(
  creatorId: string,
  sessionId: string,
): Promise<{ ok: boolean; session_id: string; status: string }> {
  return api.post(`${BASE}/${creatorId}/sessions/${sessionId}/stop`, {});
}

export async function scheduleBroadcast(
  creatorId: string,
  req: BroadcastScheduleReq,
): Promise<{ ok: boolean; session_id: string; status: string; title: string }> {
  return api.post(`${BASE}/${creatorId}/sessions/schedule`, req);
}

// -- Moderator management --

export async function registerModerator(
  creatorId: string,
  sessionId: string,
): Promise<BroadcastModeratorOut> {
  return api.post(`${BASE}/${creatorId}/sessions/${sessionId}/moderator/register`, {});
}

export async function listModerators(
  creatorId: string,
  sessionId: string,
): Promise<BroadcastModeratorOut[]> {
  return api.get(`${BASE}/${creatorId}/sessions/${sessionId}/moderators`);
}

export async function listBans(
  creatorId: string,
  sessionId: string,
): Promise<BroadcastBanOut[]> {
  return api.get(`${BASE}/${creatorId}/sessions/${sessionId}/bans`);
}

export async function getModerationLog(
  creatorId: string,
  sessionId: string,
  limit = 100,
): Promise<BroadcastModerationLogEntry[]> {
  return api.get(`${BASE}/${creatorId}/sessions/${sessionId}/moderation-log?limit=${limit}`);
}
